### PR TITLE
(main-next): Move main/next automation from github workflow to build-cli

### PR DESCRIPTION
**Pull request by sonalideshpandemsft**
As a result of this modification, the main/next PRs will continue to open automatically, while offering several advantages:

- Eliminates the need for Azure Queue usage.
- Enhances code readability and simplifies debugging. Currently, troubleshooting and fixing issues in the GitHub workflow is challenging.
- Ensures that the PRs are appropriately assigned to the correct owner.
- Open single PRs with conflicting commit and assign to the right owner.

Here's an overview of the implementation details:

image

fyi: `findConflictIndex` function is renamed to `hasConflict`. 

### Follow-ups

1. Load description from markdown files
2. Add docs for functions added

### CODE REVIEW

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 1]
```diff
@@ -21,6 +21,16 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
 			char: "a",
 			required: true,
 		}),
+		owner: Flags.string({
+			description: "Owner name",
+			char: "o",
+			required: true,
+		}),
+		repo: Flags.string({
+			description: "Repsoitory name",
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:30]:
```suggestion
			description: "Repository name",
```

File: build-tools/packages/build-cli/docs/generate.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ Generate commands are used to create/update code, docs, readmes, etc.
 * [`flub generate buildVersion`](#flub-generate-buildversion)
 * [`flub generate bundleStats`](#flub-generate-bundlestats)
 * [`flub generate packageJson`](#flub-generate-packagejson)
+* [`flub generate readme`](#flub-generate-readme)
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/docs/generate.md:9]: Unrelated change, I think.

**Comment by tylerbutler** [comment]
If I understand correctly, now the queue will stop on the commit with the merge conflict, and assign it to the person who created that commit? That sounds like a good improvement!

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 3]
```diff
@@ -93,89 +92,92 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+		};
+
+		// `mergeAllCommits` function will merge all commitIds in next branch
+		const mergeAllCommits = async (commitIds: string[]): Promise<void> => {
+			for (const commit of commitIds) {
+				// eslint-disable-next-line no-await-in-loop
+				const response = await gitRepo.merge(commit);
+				this.log(`Merging commit ${commit}: ${response}`);
+			}
+		};
```
[the remark this answers is not in the document]
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts:132]: yeah 😢 

we would have to cherry-pick commits in that case. what do you think?

[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts]: Let's say we have five commits, and commit 3 - the one right in the middle - has a conflict. I think you're saying that this PR will create 2 commits - 1 for commit 1 and another for commit 2. But can't we only execute the merge for commit 2?

But maybe I am really confused... We're still going to open a PR from commit 2 to next, right? And that PR is what will really get merged, not what the tools does locally?

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 4]
```diff
@@ -54,6 +55,48 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+	 * `cherryPickAllCommits` function will merge all commitIds in next branch
+	 * we cannot use gitRepo.merge as it would create a merge commit for individual commit.
+	 */
+	async cherryPickAllCommits(commitIds: string[], gitRepo: GitRepo): Promise<void> {
+		for (const commit of commitIds) {
+			// eslint-disable-next-line no-await-in-loop
+			const response = await gitRepo.cherryPick(commit);
+			this.log(`Merging commit ${commit}: ${response}`);
+		}
+	}
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:98]: We definitely don't want to cherry-pick, because that will create duplicate commits with confusing history. I think you should merge the most recent commit in the list that didn't conflict. That will create one merge commit, and then you open the PR from that merge commit to next. Then, when merging, instead of using the GitHub UI the person doing the merge fast-forwards next to the merge commit and pushes.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts]: Yes, this will open PR which has only the first two commits and this would get merged into next.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 5]
```diff
@@ -93,89 +127,50 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
 
-			if (response === "Abort merge") {
-				break;
-			}
+		const commitId = unmergedCommitList[commitInfo.index];
 
-			commit = i;
-			i++;
-		}
+		const branchName = `${flags.source}-${flags.target}-${commitId}`;
```
[the remark this answers is not in the document]
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts:143]: Also I got a little confused between this branch and the other one (next-automation). Maybe there are clearer names?

**Review by tylerbutler — APPROVED** [review]

**Review by scottn12 — APPROVED** [review]
This looks awesome! Thanks for doing this Sonali 😄

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 6]
```diff
@@ -36,50 +42,69 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+	async hasConflicts(commitIds: string[], gitRepo: GitRepo): Promise<CommitStatus> {
+		const length = commitIds.length;
+
+		// Check the first commit separately
+		const firstCommit = commitIds[0];
+		const firstResponse = await gitRepo.canMergeWithoutConflicts(firstCommit);
+		this.log(`Response from merge check for ${commitIds[0]}: ${firstResponse}`);
+		if (firstResponse === false) {
+			return { isConflict: true, index: 0 };
+		}
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:65]: Do you need to treat the first item differently? The logic is all the same except for the return, and you can just return `i` instead of `i - 1`.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 7]
```diff
@@ -36,50 +42,69 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+			const commit = commitIds[i];
+			// eslint-disable-next-line no-await-in-loop
+			const response = await gitRepo.canMergeWithoutConflicts(commit);
+			this.log(`Response from merge check for ${commitIds[i]}: ${response}`);
+			if (response === false) {
+				return { isConflict: false, index: i - 1 };
+			}
+		}
+
+		return { isConflict: false, index: length - 1 }; // No conflicts found, return the last index
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:78]:
```suggestion
		for (let i = 1; i < length; i++) {
			const commit = commitIds[i];
			// eslint-disable-next-line no-await-in-loop
			const mergesClean = await gitRepo.canMergeWithoutConflicts(commit);
			this.log(`Response from merge check for ${commitIds[i]}: ${response}`);
			if (mergesClean === false) {
				return { isConflict: true, index: i };
			}
		}

		return { isConflict: false, index: length - 1 }; // No conflicts found, return the last index
```

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 8]
```diff
@@ -36,50 +42,69 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+	}
+
 	public async run(): Promise<void> {
 		const flags = this.flags;
 
+		const title: string = `Automation: ${flags.source} ${flags.target} integrate`;
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:84]:
```suggestion
		const title: string = `Automation: ${flags.source}-${flags.target} integrate`;
```

**Review by tylerbutler — CHANGES_REQUESTED** [review]
I think there's a bug; see my feedback.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts (hunk 6 above)
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts]: `firstResponse` is false which implies conflict exists with 0th commit. The object returned in this case is the `{ isConflict: true, index: "CONFLCITING COMMIT INDEX" }`. `isConflict` is set to true indicating that 0th commit has a conflict with next and PR could be opened till that single commit.

`mergesClean` is false which implies conflict exists with ith commit. The object returned in this case is the `{ isConflict: false, index: "NON-CONFLCITING COMMIT INDEX" }`. `isConflict` is set to false indicating that (i-1)th commit does not have a conflict with next and PR could be opened till the (i-1)th commit.

We can modify the function to avoid confusion, but we would need a bunch of if-else statements in that case: 

```ts
if (commitInfo.isConflict === true && commitInfo.index === 0) {
	commitId = unmergedCommitList[0];
} else if (commitInfo.isConflict === true && commitInfo.index !== 0) {
	commitId = unmergedCommitList[commitInfo.index--];
	commitInfo.isConflict = false;
} else {
	commitId = unmergedCommitList[commitInfo.index];
}

const branchName = `${flags.source}-${flags.target}-${commitId.slice(0, 7)}`;

await gitRepo.deleteBranch(branchToCheckConflicts);
await gitRepo.switchBranch(flags.source);
await gitRepo.createBranch(branchName);
await gitRepo.setUpstream(branchName);
await gitRepo.resetBranch(commitId);
```
We want to merge next only with `isConflict` is false.

File: build-tools/packages/build-cli/src/lib/git.ts [hunk 9]
```diff
@@ -92,13 +92,28 @@ export class Repository {
...
 			.raw("merge-base", `refs/remotes/${remote}/${branch}`, localRef);
 		return base;
 	}
 
+	/**
+	 * @param branch - The branch to compare against.
+	 * @param remote - The remote to compare against.
+	 * @param localRef - The local ref to compare against. Defaults to HEAD.
+	 * @returns The ref of the merge base between the current HEAD and the remote branch.
+	 */
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/lib/git.ts:111]:
```suggestion
	/**
	 * @param ref1 - The first ref to compare.
	 * @param ref2 - The ref to compare against.
	 * @returns The ref of the merge base between the two refs.
	 */
```

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 10]
```diff
@@ -36,146 +45,262 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
-			str += id;
+		const prExists: boolean = await pullRequestExists(flags.auth, prTitle, this.logger);
 
-			if (str.length > 40) {
-				this.error(
-					`Unexpected string. Incorrect commit id length. Please check the commit id`,
-				);
-			}
+		if (prExists) {
+			this.error(`Open pull request exists`, { exit: -1 });
```
[the remark this answers is not in the document]
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts]: probably verbose would be better followed by exit

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 11]
```diff
@@ -36,146 +46,308 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+
 	public async run(): Promise<void> {
 		const flags = this.flags;
 
+		const prTitle: string = `Automation: ${flags.source}-${flags.target} integrate`;
+		console.log("prTitle---", prTitle);
 		const context = await this.getContext();
-		const gitRepo = context.gitRepo;
-		const prExists: boolean = await pullRequestExists(flags.auth, this.logger);
+		this.gitRepo ??= new Repository({ baseDir: context.gitRepo.resolvedRoot });
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:76]: ??= 🤩

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 12]
```diff
@@ -36,146 +46,308 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
 
-		if (prExists) {
-			this.exit(-1);
-			this.error(`Open pull request exists`);
+		if (this.gitRepo === undefined) {
+			console.log(`gitRepo undefined: ${JSON.stringify(this.gitRepo)}`);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:79]: Avoid using console.log; use the command logger instead.

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 13]
```diff
@@ -36,146 +46,308 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+	private readonly branchesToCleanup: string[] = [];
+
+	private gitRepo: Repository | undefined;
+	private remote: string | undefined;
+
 	public async run(): Promise<void> {
 		const flags = this.flags;
 
+		const prTitle: string = `Automation: ${flags.source}-${flags.target} integrate`;
+		console.log("prTitle---", prTitle);
```
[the remark this answers is not in the document]
Comment (sonalideshpandemsft) [reply build-tools/packages/build-cli/src/commands/merge/branches.ts]: removed. I was testing locally

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 14]
```diff
@@ -36,146 +46,307 @@ export default class MergeBranch extends BaseCommand<typeof MergeBranch> {
...
+		if (this.gitRepo === undefined) {
+			this.log(`gitRepo undefined: ${JSON.stringify(this.gitRepo)}`);
+			this.error("gitRepo is undefined", { exit: 1 });
 		}
 
-		const lastMergedCommit = await gitRepo.mergeBase(flags.source, flags.target);
-		this.log(
-			`${lastMergedCommit} is the last merged commit id between ${flags.source} and ${flags.target}`,
-		);
+		const repoURL = await this.gitRepo.gitClient.raw(["config", "--get", "remote.origin.url"]);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:82]: There's an easier way to do this. It's not obvious. Example:

https://github.com/microsoft/FluidFramework/blob/39dbbaaebe23918339a9ef10e709eaab09f2aab3/build-tools/packages/build-cli/src/handlers/checkFunctions.ts#L208-L212

This works because we always set the COntext's originRemotePartialUrl property to microsoft/FluidFramework:

https://github.com/microsoft/FluidFramework/blob/39dbbaaebe23918339a9ef10e709eaab09f2aab3/build-tools/packages/build-cli/src/base.ts#L120

It's ok to use the old wrapper here since you don't have to add any new code to it.

File: build-tools/packages/build-cli/src/lib/git.ts [hunk 15]
```diff
@@ -164,4 +178,25 @@ export class Repository {
...
+	 * @returns An array of all commits between the base and head commits.
+	 */
+	public async revList(baseCommit: string, headCommit: string = "HEAD"): Promise<string[]> {
+		const result = await this.git.raw("rev-list", `${baseCommit}..${headCommit}`);
+		return result
+			.split(/\r?\n/)
+			.filter((value) => value !== null && value !== undefined && value !== "");
+	}
+
+	public async canMergeWithoutConflicts(commit: string): Promise<boolean> {
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/lib/git.ts:196]: for follow up PR: add docs.

**Review by tylerbutler — APPROVED** [review]
🚀
